src/lib/modes/aead/aead.h

```diff
@@ -129,6 +134,14 @@ class BOTAN_PUBLIC_API(2,0) AEAD_Mode : public Cipher_Mode
       size_t default_nonce_length() const override { return 12; }
 
       virtual ~AEAD_Mode() = default;
+
+   protected:
+      /**
+       * Derived AEADs must implement this.
+       * If `maximum_associated_data_inputs()` returns 1 (the default), the
+       * @p idx must simply be ignored.
+       */
+      virtual void set_ad_n(size_t idx, std::span<const uint8_t> ad) = 0;
```

Comment:
The overload thing is a pain. I have no idea why C++ does that - this is why for example the `HashFunction`s have an `update` call that delegates to the virtual `add_data`, instead of the `update` directly being the virtual.

In this specific case I think what we could do is keep `set_associated_data_n` as the virtual entry point for setting the AD, and have it take a `span`. This function is new (only added in 2.17.0) and probably very rarely used since it only is useful for SIV at the moment, so it's probably fine to just change the signature and not provide any overloads.

Reply:
Do we need to mention that in the migration guide?